boto3utils/s3.py

```diff
@@ -29,22 +29,28 @@ def __init__(self, session=None, requester_pays=False):
 
     @classmethod
     def urlparse(cls, url):
-        """ Split S3 URL into bucket, key, filename """
+        """ Split S3 URL into bucket, key, filename, query_params """
         _url = deepcopy(url)
-        if url[0:5] == 'https':
+
+        if url.startswith("https"):
             _url = cls.https_to_s3(url)
-        if _url[0:5] != 's3://':
-            raise Exception('Invalid S3 url %s' % _url)
 
-        url_obj = _url.replace('s3://', '').split('/')
+        if not url.startswith("s3://"):
+            raise Exception(f"Invalid S3 url {_url}")
+
+        parsed = urlparse(_url)
+        query_params = parse_qs(parsed.query or "")
 
-        # remove empty items
-        url_obj = list(filter(lambda x: x, url_obj))
+        # now fix the "list-of-1" to be a straight string
+        for key in query_params:
+            if len(query_params[key]) == 1:
+                query_params[key] = query_params[key][0]
 
         return {
-            'bucket': url_obj[0],
-            'key': '/'.join(url_obj[1:]),
-            'filename': url_obj[-1] if len(url_obj) > 1 else ''
+            "bucket": parsed.netloc,
+            "key": parsed.path.removeprefix("/"),
+            "filename": os.path.basename(parsed.path),
+            "query_params": query_params,
```

Comment:
'query_params' seems a bit redundant, any problem with just calling this "parameters", or even just "params"?

Reply:
nope.  will update to `parameters` (everything else in this hash is spelled out... might as well spell parameters too)